upkie/cpp/actuation/BulletInterface.cpp

```diff
@@ -353,6 +361,33 @@ void BulletInterface::read_joint_sensors() {
     result.torque = sensor_state.m_jointMotorTorque;
   }
 }
+void BulletInterface::get_nominal_masses() {
+  const int nb_links = bullet_.getNumJoints(robot_);
+  b3DynamicsInfo info;
+  for (int link_id = 0; link_id < nb_links; ++link_id) {
+    bullet_.getDynamicsInfo(robot_, link_id, &info);
+    nominal_masses[link_id] = info.m_mass;
+    for (int i = 0; i < 3; ++i) {
+      nominal_inertia[link_id][i] = info.m_localInertialDiagonal[i];
+    }
+  }
+}
+void BulletInterface::randomize_masses() {
+  for (const auto& link_id : nominal_masses) {
+    std::default_random_engine generator;
+    std::uniform_real_distribution<double> distribution(
+        -mass_randomization_epsilon_, mass_randomization_epsilon_);
+    double epsilon = distribution(generator);
+    RobotSimulatorChangeDynamicsArgs change_dyn_args;
+    change_dyn_args.m_mass = nominal_masses[link_id.first] * (1 + epsilon);
+    for (int i = 0; i < 3; ++i) {
+      change_dyn_args.m_localInertiaDiagonal[i] =
+          nominal_inertia[link_id.first][i] * (1 + epsilon);
```

Comment:
If we do this the inertia matrix may not be consistent with the new total mass. The inertia matrix is a 3x3 matrix with structural properties, like positive semi-definitess and *triangular inequalities* (see for instance the intro of [Identification of Fully Physical Consistent Inertial Parameters using Optimization on Manifolds](https://arxiv.org/pdf/1610.08703)).

Here the three numbers Bullet stores are the principal moments of inertia (beware: they should be positive), but when we change the mass distribution these moments will change as well. As a matter of fact, not only the moments will change, but in general the principal axes may rotate as well. You can check for instance [this note on inertia tensors](https://ocw.mit.edu/courses/16-07-dynamics-fall-2009/dd277ec654440f4c2b5b07d6c286c3fd_MIT16_07F09_Lec26.pdf) for further details.

Can we find a smart way to rescale these principal moments while being consistent with the distribution shift $\int {\mathrm{d}m} = \varepsilon$?

Reply:
(On a side code note, if the four randomized values were independent, you should sample four epsilons rather than re-using `epsilon` four times here?)